src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py

```diff
@@ -0,0 +1,803 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import PIL
+import torch
+from transformers import (
+    AutoProcessor,
+    BlipForConditionalGeneration,
+    CLIPFeatureExtractor,
+    CLIPTextModel,
+    CLIPTokenizer,
+)
+
+from ...models import AutoencoderKL, UNet2DConditionModel
+from ...models.cross_attention import CrossAttention
+from ...schedulers import KarrasDiffusionSchedulers
+from ...utils import is_accelerate_available, logging, randn_tensor
+from ..pipeline_utils import DiffusionPipeline
+from . import StableDiffusionPipelineOutput
+from .safety_checker import StableDiffusionSafetyChecker
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+def generate_caption(image, captioner, processor, return_image=True):
+    """Generates caption for a given image."""
+    inputs = processor(images=image, return_tensors="pt")
+    outputs = captioner.generate(inputs)
+    caption = processor.batch_deocde(outputs, skip_special_tokens=True)[0]
+    if return_image:
+        return caption, inputs["pixel_values"]
+    else:
+        return caption
+
+
+def prepare_unet(unet: UNet2DConditionModel):
+    """Modifies the UNet (`unet`) to perform Pix2Pix Zero optimizations."""
+    pix2pix_zero_attn_procs = {}
+    for name in unet.attn_processors.keys():
+        module_name = name.replace(".processor", "")
+        module = unet.get_submodule(module_name)
+        if "attn2" in name:
+            pix2pix_zero_attn_procs[name] = Pix2PixZeroCrossAttnProcessor(is_pix2pix_zero=True)
+            module.requires_grad_(True)
+        else:
+            pix2pix_zero_attn_procs[name] = Pix2PixZeroCrossAttnProcessor(is_pix2pix_zero=False)
+            module.requires_grad_(False)
+
+    unet.set_attn_processor(pix2pix_zero_attn_procs)
+    return unet
+
+
+def construct_direction(source_embedding_path: str, target_embedding_path: str):
+    """Constructs the edit direction to steer the image generation process semantically."""
+    embs_source = torch.load(source_embedding_path)
+    embs_target = torch.load(target_embedding_path)
+    return (embs_target.mean(0) - embs_source.mean(0)).unsqueeze(0)
+
+
+class Pix2PixZeroL2Loss:
+    def __init__(self):
+        self.loss = 0.0
+
+    def compute_loss(self, predictions, targets):
+        self.loss += ((predictions - targets) ** 2).sum((1, 2)).mean(0)
```

Comment:
Reasoning behind `compute_loss()`:

This way, users can pass arbitrary loss functions on the regression space (Smooth L1, for example). I think design-wise it's a tad bit more flexible.